Storage/DiskImageStore.swift

```diff
@@ -84,18 +84,22 @@ open class DiskImageStore {
     }
 
     /// Clears all images from the cache, excluding the given set of keys.
-    open func clearExcluding(_ keys: Set<String>) {
-        let keysToDelete = self.keys.subtracting(keys)
+    open func clearExcluding(_ keys: Set<String>) -> Success {
+        return deferDispatchAsync(queue) { () -> Success in
+            let keysToDelete = self.keys.subtracting(keys)
 
-        for key in keysToDelete {
-            let url = URL(fileURLWithPath: filesDir).appendingPathComponent(key)
-            do {
-                try FileManager.default.removeItem(at: url)
-            } catch {
-                log.warning("Failed to remove DiskImageStore item at \(url.absoluteString): \(error)")
+            for key in keysToDelete {
+                let url = URL(fileURLWithPath: self.filesDir).appendingPathComponent(key)
+                do {
+                    try FileManager.default.removeItem(at: url)
```

Comment:
If you want to avoid certain kinds of problems, consider using a barrier. A barrier task will temporarily turn a concurrent queue into a serial queue, which might make sense for single-writer multiple-reader.